src/main/scala/util/HeterogeneousBag.scala

```diff
@@ -6,12 +6,34 @@ import Chisel._
 import chisel3.Record
 import scala.collection.immutable.ListMap
 
-final case class HeterogeneousBag[T <: Data](elts: Seq[T]) extends Record with collection.IndexedSeq[T] {
+final case class HeterogeneousBag[T <: Data](elts: Seq[T], eltNames: Option[Seq[String]] = None)
+    extends Record with collection.IndexedSeq[T] {
+
+  eltNames.foreach{ n =>
+    require(elts.size == n.size,
+    s"If eltNames are specified, they must be specified for all elts " +
+      s" (eltNames.size = ${n.size} but elts.size = ${elts.size})")
+    require(n.distinct.size == n.size, "If eltNames are specified they must all be specified & unique")
+  }
+
+  private def nameFromIndex(i: Int): String = eltNames.map(_(i)).getOrElse(i.toString)
+
+  private val nameToElts: Map[String, T] = ListMap(elts.zipWithIndex.map { case (elt, i) =>
+    val name = nameFromIndex(i)
+    name -> elt
```

Comment:
nit: no need for a `val` you only use once. i would just `(nameFromIndex(i), elt)` like you did below... and actually i would just use `nameToElts` there directly

Reply:
yeah this was probably some intermediate debugging step, will clean it up

Reply:
this was all reverted